### PR TITLE
Add ability to configure machines inside stack definition [CHE-2515]

### PR DESCRIPTION
**Pull request by olexii4**
### What does this PR do?

Add ability to configure machines inside stack definition
#2515

https://github.com/eclipse/che/issues/2727

### CODE REVIEW

**Review by akurinnoy — APPROVED** [review]

**Comment by TylerJewell** [comment]
@slemeur - are we calling them "environments" or should that label be changed to "Runtimes"?

**Review by benoitf — APPROVED** [review]

File: dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ */
+'use strict';
+import {ListCommandsController} from "../list-commands.controller";
+
+/**
+ * @ngdoc controller
+ * @name list.environment.variables.controller:EditCommandDialogController
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts]: workspace.details.controller instead list.environment.variables

File: dashboard/src/app/workspaces/workspace-details/list-commands/add-command-dialog/add-command-dialog.controller.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ */
+'use strict';
+import {ListCommandsController} from "../list-commands.controller";
+
+/**
+ * @ngdoc controller
+ * @name list.environment.variables.controller:AddCommandDialogController
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/list-commands/add-command-dialog/add-command-dialog.controller.ts]: workspace.details.controller list.environment.variables

File: dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    this.commands.forEach((command) => {
+      if (this.commandsSelectedStatus[command.name]) {
+        this.selectedCommandsNumber++;
+      } else {
+        this.isBulkChecked = false;
+      }
+    });
+  }
+
+  changePortSelection(name): void {
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts]: changePortSelection -> changeCommandSelection

File: dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      return;
+    }
+    this.selectAllPorts();
+    this.isBulkChecked = true;
+  }
+
+  /**
+   * Check all ports in list
+   */
+  selectAllPorts(): void {
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts]: selectAllCommands

File: dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    this.selectedCommandsNumber = this.commands.length;
+    this.commands.forEach((command) => {
+      this.commandsSelectedStatus[command.name] = true;
+    })
+  }
+
+  /**
+   * Uncheck all ports in list
+   */
+  deselectAllPorts(): void {
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts]: deselectAllCommands

**Review by ashumilova — APPROVED** [review]

File: dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts (hunk 1 above)
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts]: list.commands.controller  instead  list.environment.variables.controller ?

Comment (ashumilova) [reply dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts]: Ok

File: dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts (hunk 5 above)
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts]: done

File: dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts (hunk 4 above)
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts]: done

File: dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts (hunk 3 above)
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/list-commands/list-commands.controller.ts]: done

File: dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts (hunk 1 above)
Comment (olexii4) [reply dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts]: done

**Review by akurinnoy — APPROVED** [review]

File: dashboard/src/app/stacks/stack-details/list-components/add-component-dialog/add-component-dialog.html [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+-->
+<che-popup che-title="Add a new component" che-on-close="addComponentDialogController.hide()">
+  <ng-form name="addComponentForm" class="add-component-form">
+    <div class="form-input-fields">
+      <che-input type="text"
+                 che-form="addComponentForm"
+                 che-label-name="Name"
+                 che-name="name"
+                 che-place-holder="enter component's name"
```
Comment (ashumilova) [inline dashboard/src/app/stacks/stack-details/list-components/add-component-dialog/add-component-dialog.html]: enter component's name - Enter component's name

File: dashboard/src/app/stacks/stack-details/list-components/add-component-dialog/add-component-dialog.html [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                 ng-maxlength="256"
+                 required>
+        <div ng-message="required">Component's name is required.</div>
+        <div ng-message="customValidator">This component's name is already in use.</div>
+        <div ng-message="maxlength">Component's name should be less than 256 characters long.</div>
+      </che-input>
+      <che-input che-form="addComponentForm"
+                 che-label-name="Version"
+                 che-name="version"
+                 che-place-holder="enter component's version"
```
Comment (ashumilova) [inline dashboard/src/app/stacks/stack-details/list-components/add-component-dialog/add-component-dialog.html]: enter component's version -> Enter component's version

File: dashboard/src/app/stacks/stack-details/list-components/edit-component-dialog/edit-component-dialog.controller.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+'use strict';
+import {ListComponentsController} from "../list-components.controller";
+
+/**
+ * @ngdoc controller
+ * @name list.components.controller:EditComponentDialogController
+ * @description This class is handling the controller for a dialog box about editing the stack's component.
+ * @author Oleksii Orel
+ */
+export class EditComponentDialogController {
```
Comment (ashumilova) [inline dashboard/src/app/stacks/stack-details/list-components/edit-component-dialog/edit-component-dialog.controller.ts:20]: we can use same widget for adding new and editing, cause they are the same.

File: dashboard/src/app/stacks/stack-details/list-components/list-components.controller.ts [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    let component: any = {};
+    component.name = name;
+    component.version = version ? version : '---';
+    this.components.push(component);
+    this.updateSelectedStatus();
+    this.componentsOnChange();
+  }
+
+  /**
+   * Update the component
```
Comment (ashumilova) [inline dashboard/src/app/stacks/stack-details/list-components/list-components.controller.ts:114]: update doc header with params

File: dashboard/src/app/workspaces/workspace-details/environments/add-machine-dialog/add-machine-dialog.controller.ts [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    }
+    this.machineName = keys[0];
+    this.machineRAM = this.machineRecipe[keys[0]]['mem_limit'];
+  }
+
+  /**
+   * Update machine RAM.
+   */
+  updateMachineRAM() {
+    this.machineRecipe[Object.keys(this.machineRecipe)[0]]['mem_limit'] = this.machineRAM;
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/environments/add-machine-dialog/add-machine-dialog.controller.ts:114]: use EnvironmentManager when perform any actions or transformations on machine and environment.

File: dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+'use strict';
+import {ListCommandsController} from "../list-commands.controller";
+
+/**
+ * @ngdoc controller
+ * @name list.commands.controller:EditCommandDialogController
+ * @description This class is handling the controller for a dialog box about editing the workspace's command.
+ * @author Oleksii Orel
+ */
+export class EditCommandDialogController {
```
Comment (ashumilova) [inline dashboard/src/app/workspaces/workspace-details/list-commands/edit-command-dialog/edit-command-dialog.controller.ts:20]: the same as add command dialog

File: dashboard/src/components/api/che-stack.factory.ts [hunk 12]
```diff
@@ -43,25 +58,89 @@ export class CheStack {
...
+   * Gets stack template
+   * @returns {stack}
+   */
+  getStackTemplate(): any {
+    let stack: any = {
+      'name': 'New Stack',
+      'description': 'New Java Stack',
+      'scope': 'general',
+      'tags': [
+        'Java 1.8'
```
Comment (ashumilova) [inline dashboard/src/components/api/che-stack.factory.ts:71]: this tag is not displayed in tags section on creation

Comment (olexii4) [reply dashboard/src/components/api/che-stack.factory.ts:71]: it have fixed  https://github.com/eclipse/che/pull/2734/commits/74ccff2e8fba1539cfaffe43c503dc268e367504#diff-f32fbb5cbd6b9b53264cc71d6a04ef37R78

File: dashboard/src/app/stacks/stack-details/list-components/edit-component-dialog/edit-component-dialog.controller.ts (hunk 8 above)
Comment (olexii4) [reply dashboard/src/app/stacks/stack-details/list-components/edit-component-dialog/edit-component-dialog.controller.ts:20]: done

**Pull request merged**